ext/dom/tests/DOMDocument_createAttributeNS_prefix_conflict.phpt

```diff
@@ -0,0 +1,82 @@
+
+// Adheres to DOM Core Level 3
+
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns1', 'hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns2', 'hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns3', 'hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns4', 'hello'))?->namespaceURI);
+echo $doc->saveXML(), "\n";
+
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns1', 'foo:hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns2', 'foo:hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns3', 'foo:hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns4', 'foo:hello'))?->namespaceURI);
+echo $doc->saveXML(), "\n";
+
+$doc = new DOMDocument();
+$doc->appendChild($doc->createElement('container'));
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns1', 'foo:hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNode($doc->createAttributeNS('http://php.net/ns1', 'hello'))?->namespaceURI);
+echo $doc->saveXML(), "\n";
+
+echo "--- setAttributeNodeNS variation ---\n";
+
+var_dump($doc->documentElement->setAttributeNodeNS($doc->createAttributeNS('http://php.net/ns1', 'hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNodeNS($doc->createAttributeNS('http://php.net/ns2', 'hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNodeNS($doc->createAttributeNS('http://php.net/ns3', 'hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNodeNS($doc->createAttributeNS('http://php.net/ns4', 'hello'))?->namespaceURI);
+echo $doc->saveXML(), "\n";
+
+var_dump($doc->documentElement->setAttributeNodeNS($doc->createAttributeNS('http://php.net/ns1', 'foo:hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNodeNS($doc->createAttributeNS('http://php.net/ns2', 'foo:hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNodeNS($doc->createAttributeNS('http://php.net/ns3', 'foo:hello'))?->namespaceURI);
+var_dump($doc->documentElement->setAttributeNodeNS($doc->createAttributeNS('http://php.net/ns4', 'foo:hello'))?->namespaceURI);
+echo $doc->saveXML(), "\n";
+
+?>
+--EXPECT--
+--- setAttributeNode variation (you shouldn't do this, but we test this anyway) ---
+NULL
+string(18) "http://php.net/ns1"
+string(18) "http://php.net/ns2"
+string(18) "http://php.net/ns3"
+<?xml version="1.0"?>
+<container xmlns="http://php.net/ns1" xmlns:default="http://php.net/ns2" xmlns:default1="http://php.net/ns3" xmlns:default2="http://php.net/ns4" default2:hello=""/>
```

Comment:
For the record, this is what firefox does (formatting by me):

```xml
'<container xmlns="http://www.w3.org/1999/xhtml" 
   a0:hello="" xmlns:a0="https://php.net/ns1"
   a1:hello="" xmlns:a1="https://php.net/ns2"
   a2:hello="" xmlns:a2="https://php.net/ns3"
   a3:hello="" xmlns:a3="https://php.net/ns4">
</container>'
```

Reply:
First of all thanks for checking!

> This test is confusing me, as it claims to test createAttributeNS but it mainly tests setAttributeNodeNS.

All subtests test the behaviour of `createAttributeNS`, it was there that there was an implementation bug so that's why I named the test like this. The first half tests the behaviour with `setAttributeNode` (no NS) and the second half `setAttributeNodeNS`.

> The W3C-Spec (https://www.w3.org/TR/DOM-Level-2-Core/core.html) for setAttributeNodeNS says: "If an attribute with that local name and that namespace URI is already present in the element, it is replaced by the new one." This is not the case for your example, as the NS differes, so it should return null for all cases. As the NS is different, I'd expect all 4 attributes to be set.

Note that the first half tests setAttributeNode, not setAttributeNodeNS: https://www.w3.org/TR/2004/REC-DOM-Level-3-Core-20040407/core.html#ID-887236154
However I need to recheck if I did the nodeName test correctly for all cases, I'll do that tomorrow.

> The <container>-Node was created without a namespace. Now, when re-parsing this, the node would be in https://php.net/ns1. That seems wrong.

Good catch! Unfortunately it looks like this has been a bug for a long time :(: https://3v4l.org/kid5o
Fixing it in this PR is appropriate, so I'll do that.

> If your implementation were correct, the replaced attribute's namespace should be dropped, because now they are redundant.

I don't think this is required by spec right?

Reply:
> 
> > The W3C-Spec (https://www.w3.org/TR/DOM-Level-2-Core/core.html) for setAttributeNodeNS says: "If an attribute with that local name and that namespace URI is already present in the element, it is replaced by the new one." This is not the case for your example, as the NS differes, so it should return null for all cases. As the NS is different, I'd expect all 4 attributes to be set.
> 
> Note that the first half tests setAttributeNode, not setAttributeNodeNS: https://www.w3.org/TR/2004/REC-DOM-Level-3-Core-20040407/core.html#ID-887236154 However I need to recheck if I did the nodeName test correctly for all cases, I'll do that tomorrow.

I did mention that these longish tests with the expects way later are confusing ;)  You're of course right, I missed the NS not being used when scrolling back. I still believe mixing DOM1 and DOM2+ APIs should just be forbidden ;) I know that's not according to the Spec though...

Then I retract my previous statement regarding the `setAttributeNode` behavior: The result makes sense, given the nodeName is identical and the non-NS-aware API should probably just ignore the Namespace.

Which would lead to the question, whether those xmlns-declarations should be there at all: _To add a new attribute node with a qualified name and namespace URI, use the setAttributeNodeNS method._  Wouldn't that imply they should be ignored?

> 
> > If your implementation were correct, the replaced attribute's namespace should be dropped, because now they are redundant.
> 
> I don't think this is required by spec right?

Not as far as i can tell. Leaving things laying around is still a bit messy ;)

Reply:
> I did mention that these longish tests with the expects way later are confusing ;)

Sorry, you're right. I'm going to split this test into two files, I should've done so from the start.

> I still believe mixing DOM1 and DOM2+ APIs should just be forbidden ;) I know that's not according to the Spec though...

I agree, it's ugly.

> Which would lead to the question, whether those xmlns-declarations should be there at all: To add a new attribute node with a qualified name and namespace URI, use the setAttributeNodeNS method. Wouldn't that imply they should be ignored?

HMM. Good question. The fact that this so vaguely worded certainly doesn't help... I'm going to check in detail tomorrow.

> Not as far as i can tell. Leaving things laying around is still a bit messy ;)

True. I'm going to fix this separately in a follow-up, as this is not the only place where it gets messy. I want to get the output correct first :)